clang/tools/driver/cc1_main.cpp

```diff
@@ -727,6 +730,12 @@ Optional<int> CompileJobCache::replayCachedResult(CompilerInstance &Clang,
       Contents = ContentsStorage;
     } else if (JustComputedResult) {
       continue;
+    } else if (OutKind == OutputKind::Dependencies) {
+      llvm::raw_svector_ostream OS(ContentsStorage);
+      if (auto E = CASDependencyCollector::replay(
+              Clang.getDependencyOutputOpts(), *CAS, BytesRef, OS))
```

Comment:
What do you think about instead of writing out the dependency file directly that `CASDependencyCollector::replay()` would pass the paths to the `DependencyCollector`s that were registered with the `CompilerInstance`?
It would allow some flexibility (work with any kind of `DependencyCollector` that may have been registered) and I think it simplifies a bit since `CASDependencyCollector ` would not need to use a `CASOutputBackend`.

Essentially instead of aiming to record a dependency file in `CASOutputs` as another `OutputKind`, we put the dependencies in a separate special node (like we do with "stderr"), then "replay" and allow the normal `DependencyCollector` to write out the file.

Reply:
The only other `DependencyCollector` that can be setup by the `CompilerInstance` itself is `ModuleDependencyCollector` and it doesn't have the same set of inputs as the `DependencyFileGenerator`.  We   explicitly scrape certain files like all the pch files from the ImplicitPCHInclude directory and feed them into `ModuleDependencyCollector` using API that are not part of the main `DependencyCollector` interface and are not passed to other collectors.  The `DependencyFileGenerator` on the other hand supports `ExtraDeps` from the output options, so it is also adding dependencies that are not part of the callbacks themselves.

We could possibly support both of these from a single collector, but it would basically require explicit handling for each collector's special behaviour and we would need to replay each of them separately as well.  I would argue unless there is a compelling reason to do this we should keep them separate.  Do we even care about `ModuleDependencyCollector` during cache replay? Maybe not.

I also thought about only handling the dependencies that would be seen in the main collector API callbacks, and re-initialize the individual collectors before replaying to add the "additional" paths that are special for each collector. This has other issues though:
* Initializaing the `ModuleDependencyCollector` requires a PP, and ASTReader, and it can read AST files. This seems quite heavy for a cache replay.
* The `ModuleDependencyCollector` is an explicit field in the CompilerInstance, not just in the list of abstract listeners.  Nothing would prevent some random code from calling its non-collector API in the future.  So then we're back to needing to actually model this specially.

Reply:
Sounds more complicated than I hoped 😅 I'm fine as-is and perhaps re-visit later; the important bit is to make the compilation key independent of output paths related to dependencies.